File: web/src/utils/skuCode.ts
/** 规格编码：仅字母与数字，最长 64 位 */
export const SKU_CODE_MAX_LEN = 64
export const SKU_CODE_RANDOM_LEN = 8
export const SKU_CODE_PATTERN = /^[A-Za-z0-9]+$/

/** 随机编码字符集（大写字母 + 数字，如 I2EKY7G8） */
const RANDOM_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

export function sanitizeSkuCode(raw: string): string {
  return raw.replace(/[^A-Za-z0-9]/g, '').slice(0, SKU_CODE_MAX_LEN)
}

export function isValidSkuCode(code: string): boolean {
  const c = code.trim()
  return c.length > 0 && c.length <= SKU_CODE_MAX_LEN && SKU_CODE_PATTERN.test(c)
}

function randomSkuCodeString(): string {
  let code = ''
  for (let i = 0; i < SKU_CODE_RANDOM_LEN; i++) {
    code += RANDOM_CHARS[Math.floor(Math.random() * RANDOM_CHARS.length)]
  }
  return code
}

/** 生成 8 位随机规格编码，确保在本批 used 集合内不重复 */
export function nextSkuCode(used: Set<string>): string {
  for (let attempt = 0; attempt < 10000; attempt++) {
    const candidate = randomSkuCodeString()
    if (!used.has(candidate)) {
      used.add(candidate)
      return candidate
    }
  }
  throw new Error('无法生成唯一规格编码，请稍后重试')
}

File: internal/pkg/util/sku_code.go
package util

import (
	"math/rand"
	"regexp"
	"strings"
	"time"
)

const SkuCodeMaxLen = 64

var skuCodePattern = regexp.MustCompile(`^[A-Za-z0-9]+$`)

// SanitizeSkuCode 只保留字母与数字
func SanitizeSkuCode(raw string) string {
	var b strings.Builder
	for _, r := range raw {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
		}
		if b.Len() >= SkuCodeMaxLen {
			break
		}
	}
	return b.String()
}

// IsValidSkuCode 校验规格编码格式
func IsValidSkuCode(code string) bool {
	code = strings.TrimSpace(code)
	if code == "" || len(code) > SkuCodeMaxLen {
		return false
	}
	return skuCodePattern.MatchString(code)
}

const skuCodeRandomLen = 8

var skuCodeRandomChars = []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")

// NextRandomSkuCode 生成随机规格编码，确保在 used 内唯一
func NextRandomSkuCode(used map[string]struct{}) string {
	for attempt := 0; attempt < 10000; attempt++ {
		b := make([]byte, skuCodeRandomLen)
		for i := range b {
			b[i] = skuCodeRandomChars[rand.Intn(len(skuCodeRandomChars))]
		}
		code := string(b)
		if _, dup := used[code]; dup {
			continue
		}
		used[code] = struct{}{}
		return code
	}
	fallback := SanitizeSkuCode(time.Now().Format("150405999"))
	if _, dup := used[fallback]; !dup {
		used[fallback] = struct{}{}
	}
	return fallback
}
